### ISSUE TITLE
Simple way to extract residuals?

### ISSUE THREAD

**Issue by multimeric** [felix-clark/ndarray-glm#19]
I note that there is no `residuals` field or method on the `Fit` object. I guess I could `fit.expectation(x)` (which by the way is not documented due to #18) and subtract `y`, but it might be a bit more user friendly to make a method for this. In particular you could provide the option to output standardised residuals, which is actually what I'm trying to calculate here.

**Comment by multimeric** [comment]
Oops, turns out this is called `errors()`, and is already implemented.

**Comment by multimeric** [comment]
Although I think these are better named "residuals", because it's a more precise term that indicates that we can't observe the true value of y.

**Comment by felix-clark** [comment]
Yes, I believe the `errors()` function is meant to check the difference between the fitted model's prediction and the true y values for an arbitrary dataset, so you can compute the residuals by passing in the training data to `errors()`. But there should be a cleaner way of getting the fit residuals directly.

Perhaps `residuals_raw()` for (y - E[y|x]) and `residuals_std()` for (y - E[y|x])/sqrt(Var[y|x]) ?

**Comment by felix-clark** [comment]
I think `residuals_response()` for the difference between value and prediction, `residuals_pearson()` for the standardized residuals, and `residuals_deviance()` for the likelihood-based version is more in line with typical GLM terminology. I do want to refresh on studentized residuals to see if/how that has a place in this interface.

**Comment by multimeric** [comment]
Yep, that would make for a good API. By the way since we're on this topic I wonder if you should provide either a field or a function for the residual sum of squares, and also the related total sum of squares, and regression sum of squares? Happy to write a PR, I'm just not sure what you've already started.

**Comment by felix-clark** [comment]
The response and standardized residuals have been implemented in 06ec0b6. I added a hook for the deviance residuals as well although it's currently unimplemented, as it will require refactoring some of the likelihood calculations.

A function for residual sum of squares probably makes sense although perhaps it should only be exposed for linear regression -- I'm not sure it's generally useful for all models. Let's address that in #21.

**Comment by felix-clark** [comment]
On a related note, I think perhaps the `Fit::errors()` function should be made private. Since it takes a `Model` argument rather than data arrays, I don't think this will be generally useful to external users right now and will probably just lead to confusion.

**Comment by multimeric** [comment]
Yep, I think that makes sense.

**Comment by felix-clark** [comment]
Deviance residuals are implemented in f4a4686.

Before closing this issue I still want to look further into studentized residuals and determine if this should be enabled by default, as an alternative or external function, or what.

**Comment by multimeric** [comment]
What do you mean by "enabled by default"?
